paperqa/rate_limiter.py

```diff
@@ -0,0 +1,382 @@
+import asyncio
+import logging
+import os
+from collections.abc import Collection
+from typing import ClassVar, Literal
+from urllib.parse import urlparse
+
+import aiohttp
+from coredis import Redis
+from limits import (
+    RateLimitItem,
+    RateLimitItemPerMinute,
+    RateLimitItemPerSecond,
+)
+from limits import (
+    parse as limit_parse,
+)
+from limits.aio.storage import MemoryStorage, RedisStorage
+from limits.aio.strategies import MovingWindowRateLimiter
+
+from .clients.crossref import CROSSREF_BASE_URL
+from .clients.semantic_scholar import SEMANTIC_SCHOLAR_BASE_URL
+
+logger = logging.getLogger(__name__)
+
+GLOBAL_RATE_LIMITER_TIMEOUT = float(os.environ.get("RATE_LIMITER_TIMOUT", "60"))
+
+# RATE_CONFIG keys are tuples, corresponding to a namespace and primary key.
+# Anything defined with MATCH_ALL variable, will match all non-matched requests for that namespace.
+# For the "get" namespace, all primary key urls will be parsed down to the domain level.
+# For example, you're trying to do a get request to "https://google.com", "google.com" will get
+# its own limit, and it will use the ("get", MATCH_ALL) for its limits.
+# machine_id is a unique identifier for the machine making the request, it's used to limit the
+# rate of requests per machine. If the primary_key is in the NO_MACHINE_ID_EXTENSIONS list, then
+# the dynamic IP of the machine will be used to limit the rate of requests, otherwise the
+# user input machine_id will be used.
```

Comment:
Wdyt of moving this directly above `RATE_CONFIG`? It's nice to keep docs next to their usage